### DISCUSSION TITLE
ParamObject on delegate function

### DISCUSSION THREAD

**Discussion by joprice** [fable-compiler/Fable#3984] [General]
When writing bindings for something like a hook that returns a function, where the returned function takes a params object, it would be convenient to be able to define a delegate that uses `ParamObject`:

```fsharp
[<ParamObject(1)>]
type NavigateFunction = delegate of to_: string * ?replace: bool * ?state: obj option

[<Import("useNavigate", "react-router")>]
let useNavigate () : NavigateFunction = nativeOnly
```

The alternative is using a type with `Emit("$0($1...)")`

```fsharp
type NavigateFunction =
    [<Emit("$0($1...)"); ParamObject(1)>]
    abstract Invoke:
      to_: string *
      ?replace: bool *
      ?state: obj option ->
        unit
```

(Maybe the second one could also be made prettier and less error prone to type with an `EmitApply` or `EmitCall` helper?)

**Comment by MangelMaxime** [comment]
Could you please share the JavaScript/TypeScript code which is trying to be mapped by this code?

**Comment by joprice** [comment]
https://unpkg.com/browse/react-router@7.0.2/dist/development/index.d.ts 

```ts
interface NavigateFunction {
  (to: To, options?: NavigateOptions): void | Promise<void>;
  (delta: number): void | Promise<void>;
 }

declare function useNavigate(): NavigateFunction;
```

Also, just found this convenient interface for looking up the types for a lib https://types.kubajastrz.com/

**Comment by MangelMaxime** [comment]
The other alternative is to not use `ParamObject` but instead something closer to the TypeScript API:

This is what Glutinum generates right now:

```fs
[<AbstractClass>]
[<Erase>]
type Exports =
    [<Import("useNavigate", "REPLACE_ME_WITH_MODULE_NAME")>]
    static member useNavigate () : NavigateFunction = nativeOnly

[<AllowNullLiteral>]
[<Interface>]
type NavigateFunction =
    [<Emit("$0($1...)")>]
    abstract member Invoke: ``to``: To * ?options: NavigateOptions -> U2<unit, JS.Promise<unit>>
    [<Emit("$0($1...)")>]
    abstract member Invoke: delta: float -> U2<unit, JS.Promise<unit>>
```

In a ideal world `NavigateOptions` would be generated as a `ParamObject` so you would end up writing:

```fs
useNavigate.Invoke("to-somewhere", NavigateOptions(
    replace = true
  )
)
```

when in TypeScript you would write

```ts
useNavigate("to-somewhere", {
  replace: true
})
```

You could perhaps optimise it even more by generating something like that

**⚠️ not tested**

```fs
[<AbstractClass>]
[<Erase>]
type Exports =
    [<Emit("$0($1...)")>]
    [<Import("useNavigate", "REPLACE_ME_WITH_MODULE_NAME")>]
    abstract member Invoke: ``to``: To * ?options: NavigateOptions -> U2<unit, JS.Promise<unit>>
    [<Emit("$0($1...)")>]
    [<Import("useNavigate", "REPLACE_ME_WITH_MODULE_NAME")>]
    abstract member Invoke: delta: float -> U2<unit, JS.Promise<unit>>
```
